contrib/devtools/interactive-test-runner.py

```diff
@@ -0,0 +1,237 @@
+
+    def find_functional_tests(self):
+        stdout = self.run_command('git', 'ls-files', '*.py', cwd=self.functional_test_root)
+        pyfiles = filter(lambda line: '/' not in line, stdout.split('\n'))
+        return list(pyfiles)
+
+    def read_test_runner(self):
+        if self.functional_test_root not in sys.path:
+            sys.path.append(self.functional_test_root)
+        from test_runner import BASE_SCRIPTS
+        return BASE_SCRIPTS
+
+    def spawn_test(self, test):
+        log_stdout = tempfile.SpooledTemporaryFile()
+        log_stderr = tempfile.SpooledTemporaryFile()
+
+        args = test.split()
+        args[0] = os.path.join(self.functional_test_root, args[0])
+        args.extend(self.passon_args)
+
+        job = subprocess.Popen(
+            args=args,
+            universal_newlines=True,
+            stdout=log_stdout,
+            stderr=log_stderr
+        )
+
+        return RunningJob(test, log_stdout, log_stderr, job)
+
+    def report_result_inline(self, test_result):
+        color, line = test_result.format(self.test_name_length)
+        missing = ' ' * max(0, self.line_length - len(line))
+        self.line_length = 0
+        print('\r%s%s' % (color(line), missing), flush=True)
+
+    def report_running(self, running_jobs):
+        def format_line(color=lambda s: s):
+            jobs = ', '.join(map(lambda j: '%s (%d)' % (color(j.test), time.time() - j.started), running_jobs))
+            return 'Running %s' % jobs
+
+        line = format_line()
+        self.line_length = max(len(line), self.line_length)
+        print('\r%s' % format_line(yellow), end='', flush=True)
+
+    def run(self, test_patterns):
+        remaining_tests = [name for name in self.tests if any([p in name for p in test_patterns])]
+        for order in self.tests_order:
+            order(remaining_tests)
+        running_jobs = []
+        test_results = []
+
+        def read(file):
+            contents = file.read().decode('utf8')
+            file.close()
+            return contents
+
+        while remaining_tests or running_jobs:
+            for job in running_jobs:
+                if job.proc.poll() is not None:
+                    test_result = TestResult(
+                        test=job.test,
+                        started=job.started,
+                        returncode=job.proc.returncode,
+                        stdout=read(job.log_stdout),
+                        stderr=read(job.log_stderr)
+                    )
+                    running_jobs.remove(job)
+                    test_results.append(test_result)
+                    self.report_result_inline(test_result)
+            while remaining_tests and len(running_jobs) < self.num_jobs:
+                test = remaining_tests.pop(0)
+                running_jobs.append(self.spawn_test(test))
+            self.report_running(running_jobs)
+            time.sleep(0.5)
+
+
+def partition(predicate, items):
```

Comment:
nit: This is a function that I always recreate at some point, and I usually call it `distill`. `partition` seems to me like it should be more generic, returning a tuple with an undefined number of partitions.

Reply:
I was inspired by http://hackage.haskell.org/package/base-4.12.0.0/docs/Data-List.html#v:partition

When I google for "partition list python" I find meaningful results:

- https://gist.github.com/mfalda/11182770
- https://stackoverflow.com/questions/12188141/partitioning-arrays-in-python-based-on-a-predicate
- https://stackoverflow.com/questions/4578590/python-equivalent-of-filter-getting-two-output-lists-i-e-partition-of-a-list

Reply:
Yep, but still using a very broad term ( https://en.wikipedia.org/wiki/Partition, https://en.wikipedia.org/wiki/Partition_of_a_set ) for partitions of just 2 subsets.

Anyway, that was just a nit, I guess the similarity with Haskell's nomenclature is enough.